Implement layer hosting for WebKit extension processes

pvollan opened the pull request:
#### b674f6c410e6fd20169eb8f8da10f143feff1073
<pre>
Implement layer hosting for WebKit extension processes
<a href="https://rdar.apple.com/119381400">rdar://119381400</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=266078">https://bugs.webkit.org/show_bug.cgi?id=266078</a>

Reviewed by Andy Estes and Brent Fulgham.

Implement layer hosting for WebKit extension processes. In this implementation, the remote context is
represented by an object of type _SEHostable. In this implementation, the _SEHostable object is created
in the GPU process. The remote context is shared by extracting the context ID from the XPC representation
of the _SEHostable object, and sent to the UI process via the WebContent process as in the former
implementation. In the UI process, a _SEHostingHandle is created from the received context ID and
attached to the new view type we create in the UI process, _SEHostingView. Fencing is implemented with
the new object type _SEHostingUpdateCoordinator. This object is created in the UI process, and shared
with the GPU process from its XPC representation, which contains a Mach send right.

* Source/WebKit/GPUProcess/media/cocoa/RemoteMediaPlayerProxyCocoa.mm:
(WebKit::RemoteMediaPlayerProxy::mediaPlayerRenderingModeChanged):
(WebKit::RemoteMediaPlayerProxy::setVideoLayerSizeFenced):
* Source/WebKit/Platform/ExtraPrivateSymbolsForTAPI.h:
* Source/WebKit/Platform/cocoa/LayerHostingContext.h:
* Source/WebKit/Platform/cocoa/LayerHostingContext.mm:
(WebKit::LayerHostingContext::createForExternalHostingProcess):
(WebKit::LayerHostingContext::~LayerHostingContext):
(WebKit::LayerHostingContext::setRootLayer):
(WebKit::LayerHostingContext::rootLayer const):
(WebKit::LayerHostingContext::contextID const):
(WebKit::LayerHostingContext::setFencePort):
(WebKit::LayerHostingContext::xpcRepresentation const):
(WebKit::LayerHostingContext::commit):
* Source/WebKit/Platform/spi/Cocoa/ExtensionKitSPI.h:
* Source/WebKit/Shared/Cocoa/ExtensionKitSoftLink.h:
* Source/WebKit/Shared/Cocoa/ExtensionKitSoftLink.mm:
* Source/WebKit/Shared/Cocoa/WKProcessExtension.mm:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(-[WKLayerHostView dealloc]):
(WebKit::VideoPresentationManagerProxy::createLayerHostViewWithID):
(WebKit::VideoPresentationManagerProxy::setVideoLayerFrame):

Canonical link: <a href="https://commits.webkit.org/272703@main">https://commits.webkit.org/272703@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/52d49aa81b66bccecb7b66b8279358e28b9940ee

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32706 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11459 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34562 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35277 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29549 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13807 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8647 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29036 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33146 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9661 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29214 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8424 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8555 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36614 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29723 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29575 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34680 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8674 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6638 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32546 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10358 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7610 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9287 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9281 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
```diff
@@ -391,25 +390,6 @@ void GPUConnectionToWebProcess::didClose(IPC::Connection& connection)
     gpuProcess().removeGPUConnectionToWebProcess(*this); // May destroy |this|.
 }
 
-#if HAVE(VISIBILITY_PROPAGATION_VIEW)
```
Actually, this should never have been in the `cpp` file at all!

brentfulgham commented on Source/WebKit/GPUProcess/media/cocoa/RemoteMediaPlayerProxyCocoa.mm:
```diff
@@ -68,12 +98,20 @@
         m_inlineLayerHostingContext = LayerHostingContext::createForExternalHostingProcess();
         IntSize presentationSize = enclosingIntRect(FloatRect(layer.frame)).size();
+#if USE(EXTENSIONKIT)
+        sendRenderingModeChangedXPCMessage(presentationSize);
```
So we now message the UIProcess, which in turn messages the WebProcess, rather than the old mechanism of direct message sends to the WCP.

Is there no way for us to directly communicate between Extensions? I guess maybe only the parent process knows all of the children and can be the exchange for relaying information.

pvollan replied:
That is a good point. Currently, we do not have a direct XPC connection between the GPU process and the WebContent process. I think we should definitely look into establishing this connection in a follow up patch, although it is not strictly required for this patch, since we can send the message via the UI process.

Thanks for reviewing!

aestes replied:
Per Arne and I chatted about this offline. I think we actually do want to send the layer hosting context over CoreIPC for IPC validation purposes. That would have the nice side effect of letting us send the message directly from GPU to WC without having to create a direct XPC connection. @beidson can speak more to how/why we should use CoreIPC here.

brentfulgham commented on Source/WebKit/Platform/cocoa/LayerHostingContext.mm:
```diff
@@ -60,13 +68,29 @@
-    }];
+    };
+#if USE(EXTENSIONKIT)
+    if ([get_SEHostableClass() respondsToSelector:@selector(createHostableWithOptions:error:)]) {
```
I think you filed a radar to move away from the dynamic `respondsToSelector` once the API is complete, right?

brentfulgham commented on Source/WebKit/UIProcess/Cocoa/GPUProcessProxyCocoa.mm:
```diff
@@ -89,6 +95,53 @@
+    if (!m_gpuProcess)
+        return false;
+
+    if (!event || xpc_get_type(event) == XPC_TYPE_ERROR)
```
Should we be logging when we receive an XPC error?

pvollan replied:
That is a good point. I will add logging in the upcoming patch.

brentfulgham commented on Source/WebKit/UIProcess/Cocoa/GPUProcessProxyCocoa.mm:
```diff
@@ -89,6 +95,53 @@
+        auto mediaPlayerIdentifier = xpc_dictionary_get_uint64(event, mediaplayerIdentifierKey);
+        auto webProcessIdentifier = xpc_dictionary_get_uint64(event, webProcessIdentifierKey);
+        RefPtr<WebProcessProxy> webProcess = WebProcessProxy::processForIdentifier(WebCore::ProcessIdentifier(webProcessIdentifier));
+        if (webProcess) {
```
Nit: You could scope `webProcess` to the `if () { ...` statement.

pvollan replied:
Done!

brentfulgham commented on Source/WebKit/WebProcess/cocoa/HandleXPCEndpointMessages.mm:
```diff
@@ -48,6 +57,21 @@ void handleXPCEndpointMessages(xpc_object_t event, const char* messageName)
+            auto identifier = xpc_dictionary_get_uint64(event.get(), mediaplayerIdentifierKey);
+            auto width = xpc_dictionary_get_uint64(event.get(), presentationSizeWidthKey);
+            auto height = xpc_dictionary_get_uint64(event.get(), presentationSizeHeightKey);
+            WeakPtr<MediaPlayerPrivateRemote> player = WebProcess::singleton().remoteMediaPlayerManager().mediaPlayerFromIdentifier(WebCore::MediaPlayerIdentifier(identifier));
```
Could this `WeakPtr` declaration just be `auto`?

pvollan replied:
Fixed!

brentfulgham commented on Source/WebKit/WebProcess/cocoa/HandleXPCEndpointMessages.mm:
```diff
@@ -48,6 +57,21 @@ void handleXPCEndpointMessages(xpc_object_t event, const char* messageName)
+        callOnMainRunLoop([event = OSObjectPtr<xpc_object_t>(event)] () {
+            auto identifier = xpc_dictionary_get_uint64(event.get(), mediaplayerIdentifierKey);
+            auto width = xpc_dictionary_get_uint64(event.get(), presentationSizeWidthKey);
+            auto height = xpc_dictionary_get_uint64(event.get(), presentationSizeHeightKey);
```
It seems like you could avoid the `width` and `height` lookups if the `player` weak pointer happens to be nullptr.

pvollan replied:
Fixed!

brentfulgham reviewed (approved):
A few minor nits, but this looks fine. We should keep an eye on performance bots to make sure this does not impact performance.

aestes commented on Source/WebKit/Platform/spi/Cocoa/ExtensionKitSPI.h:
```diff
@@ -27,8 +27,11 @@
+#import <UIKit/UIKit.h>
+
+#if __has_include(<ServiceExtensions/SEServiceManager_Private.h>)
+#import <ServiceExtensions/SEServiceManager_Private.h>
```
Is this change because `SEServiceManager_Private.h` is not imported by `ServiceExtensions_Private.h`? If so, that's a mistake; `ServiceExtensions_Private.h` should import all the private headers (I know that's not something you can fix in this PR though).

Also, doesn't `ServiceExtensions_Private.h` import things we need that `SEServiceManager_Private.h` doesn't, like `SEProcess_Private.h`?

pvollan replied:
Ah, this change was not intentional, I reverted this part.

aestes commented on Source/WebKit/Shared/Cocoa/ExtensionKitSoftLink.h:
```diff
@@ -33,5 +33,9 @@
+SOFT_LINK_CLASS_FOR_HEADER(WebKit, _SEHostable)
+SOFT_LINK_CLASS_FOR_HEADER(WebKit, _SEHostingHandle)
+SOFT_LINK_CLASS_FOR_HEADER(WebKit, _SEHostingView)
+SOFT_LINK_CLASS_FOR_HEADER(WebKit, _SEHostingUpdateCoordinator)
```
Nit: sorting (U < V)

pvollan replied:
Fixed!

aestes commented on Source/WebKit/Shared/Cocoa/ExtensionKitSoftLink.mm:
```diff
@@ -32,5 +32,9 @@
+SOFT_LINK_CLASS_FOR_SOURCE(WebKit, ServiceExtensions, _SEHostable)
+SOFT_LINK_CLASS_FOR_SOURCE(WebKit, ServiceExtensions, _SEHostingHandle)
+SOFT_LINK_CLASS_FOR_SOURCE(WebKit, ServiceExtensions, _SEHostingView)
+SOFT_LINK_CLASS_FOR_SOURCE(WebKit, ServiceExtensions, _SEHostingUpdateCoordinator)
```
Nit: sorting (U < V)

pvollan replied:
Done!

aestes commented on Source/WebKit/Platform/cocoa/LayerHostingContext.mm:
```diff
@@ -32,6 +32,14 @@
+SOFT_LINK_FRAMEWORK_OPTIONAL(ServiceExtensions);
+SOFT_LINK_CLASS_OPTIONAL(ServiceExtensions, _SEHostable);
+SOFT_LINK_CLASS_OPTIONAL(ServiceExtensions, _SEHostingUpdateCoordinator);
+#endif
```
Why aren't we importing `ExtensionKitSoftLink.h` here?

pvollan replied:
That's a good point. I got link errors when building the TestIPC target when doing that, since LayerHostingContext.mm is linked there. I think this can be resolved by moving ExtensionKitSoftLink to WebCore, since TestIPC is linking WebCore. I can address that in a follow-up patch.

Thanks for reviewing!

aestes commented on Source/WebKit/Platform/cocoa/LayerHostingContext.mm:
```diff
@@ -112,20 +127,40 @@
 {
+#if USE(EXTENSIONKIT)
+    if (m_hostable) {
+        m_hostable.get().layer = rootLayer;
```
Nit: this is fine, but our style preference in these cases is to omit the `.get()` by writing `[m_hostable setLayer:rootLayer]`.

aestes commented on Source/WebKit/Platform/cocoa/LayerHostingContext.mm:
```diff
@@ -112,20 +127,40 @@
 {
+#if USE(EXTENSIONKIT)
+    if (m_hostable)
+        return m_hostable.get().layer;
```
Nit: this is fine, but our style preference in these cases is to omit the `.get()` by writing `[m_hostable layer]`.

aestes commented on Source/WebKit/Platform/cocoa/LayerHostingContext.mm:
```diff
@@ -112,20 +127,40 @@
+    if (xpcDictionary) {
+        auto cid = xpc_dictionary_get_uint64(xpcDictionary.get(), "cid");
+        return cid;
+    }
```
Nit:
```
if (auto xpcDictionary = xpcRepresentation())
    return xpc_dictionary_get_uint64(xpcDictionary.get(), "cid");
```

aestes commented on Source/WebKit/Platform/cocoa/LayerHostingContext.mm:
```diff
@@ -176,4 +221,18 @@
+{
+    if (!m_hostable)
+        return nullptr;
+    return [m_hostable.get().handle xpcRepresentation];
```
Nit:  this is fine, but our style preference in these cases is to omit the `.get()` by writing `[[m_hostable handle] xpcRepresentation]`.

aestes commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -718,7 +736,25 @@ - (BOOL)prefersStatusBarHidden
+        auto handle = adoptNS([alloc_SEHostingHandleInstance() initFromXPCRepresentation:xpcRepresentation.get()]);
+        auto hostingView = adoptNS([[_SEHostingView alloc] init]);
+        view->_hostingView = hostingView;
+        view->_hostingView.get().handle = handle.get();
```
Nit: `[hostingView setHandle:handle.get()];`

aestes reviewed (approved):
r=me with nits mostly, although I'm a little worried about the changes to ExtensionKitSPI.h (but if everything builds then I guess it's fine). I think the commit message also needs to be updated.

pvollan committed:
Implement layer hosting for WebKit extension processes
rdar://119381400
https://bugs.webkit.org/show_bug.cgi?id=266078

Reviewed by Andy Estes and Brent Fulgham.

Implement layer hosting for WebKit extension processes. In this implementation, the remote context is
represented by an object of type _SEHostable. In this implementation, the _SEHostable object is created
in the GPU process. The remote context is shared by extracting the context ID from the XPC representation
of the _SEHostable object, and sent to the UI process via the WebContent process as in the former
implementation. In the UI process, a _SEHostingHandle is created from the received context ID and
attached to the new view type we create in the UI process, _SEHostingView. Fencing is implemented with
the new object type _SEHostingUpdateCoordinator. This object is created in the UI process, and shared
with the GPU process from its XPC representation, which contains a Mach send right.

* Source/WebKit/GPUProcess/media/cocoa/RemoteMediaPlayerProxyCocoa.mm:
(WebKit::RemoteMediaPlayerProxy::mediaPlayerRenderingModeChanged):
(WebKit::RemoteMediaPlayerProxy::setVideoLayerSizeFenced):
* Source/WebKit/Platform/ExtraPrivateSymbolsForTAPI.h:
* Source/WebKit/Platform/cocoa/LayerHostingContext.h:
* Source/WebKit/Platform/cocoa/LayerHostingContext.mm:
(WebKit::LayerHostingContext::createForExternalHostingProcess):
(WebKit::LayerHostingContext::~LayerHostingContext):
(WebKit::LayerHostingContext::setRootLayer):
(WebKit::LayerHostingContext::rootLayer const):
(WebKit::LayerHostingContext::contextID const):
(WebKit::LayerHostingContext::setFencePort):
(WebKit::LayerHostingContext::xpcRepresentation const):
(WebKit::LayerHostingContext::commit):
* Source/WebKit/Platform/spi/Cocoa/ExtensionKitSPI.h:
* Source/WebKit/Shared/Cocoa/ExtensionKitSoftLink.h:
* Source/WebKit/Shared/Cocoa/ExtensionKitSoftLink.mm:
* Source/WebKit/Shared/Cocoa/WKProcessExtension.mm:
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
(-[WKLayerHostView dealloc]):
(WebKit::VideoPresentationManagerProxy::createLayerHostViewWithID):
(WebKit::VideoPresentationManagerProxy::setVideoLayerFrame):

Canonical link: https://commits.webkit.org/272703@main

smfr commented on Source/WebKit/Platform/cocoa/LayerHostingContext.mm:
```diff
@@ -60,13 +68,20 @@
+    };
+#if USE(EXTENSIONKIT)
+    if (options.useHostable && [get_SEHostableClass() respondsToSelector:@selector(createHostableWithOptions:error:)]) {
+        layerHostingContext->m_hostable = [get_SEHostableClass() createHostableWithOptions:contextOptions error:nil];
```
This is a leak: assigning to a `RetainPtr<>` from a function that presumably returns a retained value

smfr commented on Source/WebKit/Platform/cocoa/LayerHostingContext.mm:
```diff
@@ -112,20 +127,37 @@
 {
+#if USE(EXTENSIONKIT)
+    if (auto xpcDictionary = xpcRepresentation())
+        return xpc_dictionary_get_uint64(xpcDictionary.get(), "cid");
```
This magic "cid" string should be a constant somewhere so both ends use it.

smfr commented on Source/WebKit/Platform/spi/Cocoa/ExtensionKitSPI.h:
```diff
@@ -111,4 +116,30 @@ NS_ASSUME_NONNULL_END
+@end
+
+@interface _SEHostable: NSObject
++(_SEHostable*)createHostableWithOptions:(NSDictionary*)dict error:(NSError**)error;
```
ObjC style isn't to use `create`, it would be just `hostableWithOptions:` and return an autoreleased object.

smfr commented on Source/WebKit/Platform/spi/Cocoa/ExtensionKitSPI.h:
```diff
@@ -111,4 +116,30 @@ NS_ASSUME_NONNULL_END
+@end
+
+@interface _SEHostingUpdateCoordinator : NSObject
+-(instancetype)init;
```
No need to declare `-init`

smfr commented on Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
```diff
@@ -1177,10 +1213,23 @@ - (BOOL)prefersStatusBarHidden
+    if (view && view->_hostingView) {
+        if (!view->_hostingUpdateCoordinator) {
+            auto hostingUpdateCoordinator = adoptNS([alloc_SEHostingUpdateCoordinatorInstance() init]);
+            [hostingUpdateCoordinator addHostingView:view->_hostingView.get()];
```
Do we have a retain cycle here?
